### PR TITLE
Binary Heap

### LINKED ISSUE #126: Binary Heap
The Binary Heap is commonly used for priority queues with applications in DeFi such as a DEX order book. This data structure is partially sorted, therefore only the root node is guaranteed to be in the correct order. This is particularly useful in cases where only the first entry in the queue is needed per transaction. The Binary Heap should have a generalized implementation.

### PR DESCRIPTION
**Pull request by 0xCourtney**
closes #126

### CODE REVIEW

**Comment by 0xCourtney** [comment]
The `require` statement should be changed to revert, I will do this after #139 is merged.

File: test/data/BinaryHeap.ts [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+          checkNodes(await instance['toArray()']());
+        });
+
+        it('returns true if value is added', async () => {
+          expect(await instance.callStatic['add(bytes32)'](zeroBytes32)).to.be
+            .true;
+        });
+
+        it('returns false if value has already been added', async () => {
+          await instance.callStatic['remove(bytes32)'](zeroBytes32);
```
Comment (NouDaimon) [inline test/data/BinaryHeap.ts:201]:
```suggestion
          await instance['add(bytes32)'](zeroBytes32);
```

File: test/data/BinaryHeap.ts [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+          expect(await instance.callStatic['remove(bytes32)'](zeroBytes32)).to
```
Comment (NouDaimon) [inline test/data/BinaryHeap.ts:202]:
```suggestion
          expect(await instance.callStatic['add(bytes32)'](zeroBytes32)).to
```

File: test/data/BinaryHeap.ts [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+          checkNodes(await instance['toArray()']());
+        });
+
+        it('returns true if value is added', async () => {
+          expect(await instance.callStatic['add(address)'](zeroAddress)).to.be
+            .true;
+        });
+
+        it('returns false if value has already been added', async () => {
+          await instance.callStatic['remove(address)'](zeroAddress);
```
Comment (NouDaimon) [inline test/data/BinaryHeap.ts:471]:
```suggestion
          await instance['add(address)'](zeroAddress);
```

File: test/data/BinaryHeap.ts [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+          expect(await instance.callStatic['remove(address)'](zeroAddress)).to
```
Comment (NouDaimon) [inline test/data/BinaryHeap.ts:472]:
```suggestion
          expect(await instance.callStatic['add(address)'](zeroAddress)).to
```

File: test/data/BinaryHeap.ts [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+          await instance['add(uint256)'](one);
+          checkNodes(await instance['toArray()']());
+        });
+
+        it('returns true if value is added', async () => {
+          expect(await instance.callStatic['add(uint256)'](zero)).to.be.true;
+        });
+
+        it('returns false if value has already been added', async () => {
+          await instance.callStatic['remove(uint256)'](zero);
```
Comment (NouDaimon) [inline test/data/BinaryHeap.ts:740]:
```suggestion
          await instance['add(uint256)'](zero);
```

File: test/data/BinaryHeap.ts [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+          expect(await instance.callStatic['remove(uint256)'](zero)).to.be
```
Comment (NouDaimon) [inline test/data/BinaryHeap.ts:741]:
```suggestion
          expect(await instance.callStatic['add(uint256)'](zero)).to.be
```

**Review by NouDaimon — APPROVED** [review]

File: contracts/data/BinaryHeap.sol [hunk 7]
```diff
@@ -0,0 +1,341 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.8;
+
+/**
+ * @title Binary Heap implementation
+ * @dev The data strucutre is configured as a max-heap
```
Comment (ItsNickBarry) [inline contracts/data/BinaryHeap.sol:7]:
```suggestion
 * @dev The data structure is configured as a max-heap
```

File: contracts/data/BinaryHeap.sol [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+
+pragma solidity ^0.8.8;
+
+/**
+ * @title Binary Heap implementation
+ * @dev The data structure is configured as a max-heap
+ */
+
+library BinaryHeap {
+    error BinaryHeap__IndexOutOfBounds();
```
Comment (ItsNickBarry) [inline contracts/data/BinaryHeap.sol:11]: I think we can remove this error and fall back to the default panic opcode.

**Review by ItsNickBarry — COMMENTED** [review]
@0xCourtney I made a few optimizations that you should look at.  Let me know if you agree.

File: contracts/data/BinaryHeap.sol [hunk 9]
```diff
@@ -304,7 +304,7 @@ library BinaryHeap {
         bytes32[] storage values = heap._values;
 
         unchecked {
-            uint256 left = 2 * index + 1;
+            uint256 left = (index << 1) | 1;
```
Comment (0xCourtney) [inline contracts/data/BinaryHeap.sol:301]: 🔥

File: contracts/data/BinaryHeap.sol (hunk 8 above)
Comment (0xCourtney) [reply contracts/data/BinaryHeap.sol]: https://github.com/solidstate-network/solidstate-solidity/pull/143/commits/c692368e9a52ba3a541db97b4d8fe1966192d2a6

**Review by ItsNickBarry — APPROVED** [review]
@0xCourtney I made some significant changes to the tests.  Random values are used  instead of the precalculated constants.  This doesn't improve the integrity of the tests, but they're now a bit more concise.

Merge if you're okay with these changes.

**Comment by ItsNickBarry** [comment]
Actually, I'm going to merge now so that the `data-structures` branches can be consolidated and published.  Comment if you have issues with the test updates.

**Pull request merged**